### DISCUSSION TITLE
Why doesn't this query work?

### DISCUSSION THREAD

**Discussion by nikkwong** [sqlalchemy/sqlalchemy#9786] [Usage Questions]
Hi, I have the following sqlalchemy orm code:

```
    assignment = (
        session.query(Assignment)
        .outerjoin(
            Event,
            and_(
                Assignment.assignment_id == Event.assignment_id,
                Event.event_name != "prl_step",
            ),
        )
        .where(Assignment.user_id == g.user.user_id)
        .first()
    )
```

in which I'm trying to essentially issue the following query:

```
SELECT
  app.assignments.created_at AS app_assignments_created_at,
  app.assignments.user_id AS app_assignments_user_id
FROM
  app.assignments
  LEFT OUTER JOIN app.events ON app.assignments.assignment_id = app.events.assignment_id
  AND app.events.event_name != 'prl_step'
WHERE
  app.assignments.user_id = % user_id %
```

The ORM is returning events which have `Event.event_name == prl_step` even though it shouldn't be. When I'm looking at sqlalchemy in debug mode, it does look like it's actually issuing the query correctly:

```
INFO:sqlalchemy.engine.Engine:SELECT app.assignments.created_at AS app_assignments_created_at, app.assignments.user_id AS app_assignments_user_id, app.assignments.assignment_id AS app_assignments_assignment_id, app.assignments.assignment_source_id AS app_assignments_assignment_source_id, app.assignments.begin_timestamp AS app_assignments_begin_timestamp, app.assignments.end_timestamp AS app_assignments_end_timestamp 
FROM app.assignments LEFT OUTER JOIN app.events ON app.assignments.assignment_id = app.events.assignment_id AND app.events.event_name != %(event_name_1)s 
WHERE app.assignments.user_id = %(user_id_1)s 
 LIMIT %(param_1)s
```

But the event results are still coming through in the ORM. Can someone help me understand why? Thank you.

**Comment by zzzeek** [comment]
if you OUTER JOIN from assignment to events, the rows for which the "events" table do not match the ON criteria will continue to return a row, because LEFT OUTER JOIN means "return the left side of the row and all NULLs for the right side of the row" for non matching rows.

so to use LEFT OUTER JOIN to exclude entire rows, which is not usually how I would do this, you'd have to exclude the NULLs also, something like:

```py
    assignment = (
        session.query(Assignment)
        .outerjoin(
            Event,
            and_(
                Assignment.assignment_id == Event.assignment_id,
                Event.event_name != "prl_step",
            ),
        )
        .where(Assignment.user_id == g.user.user_id).where(Event.id != None)
        .first()
    )
```

Usually the best way to exclude rows based on something happening in a related table is to use EXISTS or NOT EXISTS.  So this query I would write as:

```py
assignment = (
    session.query(Assignment)
    .where(
        ~exists().where(
            and_(
                Assignment.assignment_id == Event.assignment_id,
                Event.event_name == "prl_step",
            )
        )
    )
    .where(Assignment.user_id == g.user.user_id)
    .first()
)
```

that would give you "assignment...where NOT EXISTS (any event where name == "prl_step" and event.assignment_id==assignment.assignment_id".    these usually work better for this kind of thing.

**Reply by nikkwong** [reply]
This is so helpful! I'm sorry, but my intention was to include the assignment result, but not return the events with `Event.event_name == 'prl_step'`. So the result from the ORM should be like: 

```
{
    assignment_id: 123,
    events: [{
        event_name: 'something'
     }] // disclude event_name == 'prl_step'
}
```

Thank you so much for the response.

**Reply by zzzeek** [reply]
your query is returning a list of Assignment results.  so if you are trying to affect the contents of each Assignment.events collection you have to link the outerjoin to the collection using [contains_eager](https://docs.sqlalchemy.org/en/20/orm/queryguide/relationships.html#contains-eager).

**Reply by nikkwong** [reply]
Thank you so much, I got it! 

One odd thing about my approach is that if I chain the end of the call with `.first()` it limits the Assignment.events results to the first result. So `len(assignment.events) == 1`. I'm instead calling `.all()` which returns all of the events but then also an array of `assignments` as the return value although I'm only expecting one to match the query (see the `Assignment.id == id) line: 

```
assignments = (
    session.query(Assignment)
    .join(Assignment.events)
    .options(contains_eager(Assignment.events))
    .filter(Event.event_name != "prl_step")
    .filter(Assignment.id == id)
    .all()
)
assignment = assignments[0]
```

Not sure if that's tricky or not, but this is great for now, thank you so much!!

**Reply by nikkwong** [reply]
Well, this query doesn't work. If no events are present, it will return `None` for `assignments` even though the assignment is present. I've also tried `.outerjoin()` instead of `.join()` Is there a better way around this? Thanks :)
